luacompiler/src/lib/irgen/mod.rs

```diff
@@ -140,14 +145,68 @@ impl<'a> LuaToIR<'a> {
         }
     }
 
+    /// Compiles a local multi-assignemnt.
+    /// * `names` - the variable names
+    /// * `exprs` - the expressions that are assigned
+    fn compile_local_assignments(&mut self, names: Vec<&'a str>, exprs: Vec<&'a Node<u8>>) {
+        // example: local a, b, c, d = 1, 2
+        // compile local a = 1, local b = 2
+        for i in 0..exprs.len() {
+            // left hand-side = <namelist> and right hand-side = <explist>
+            self.compile_assignment(names[i], exprs[i], true);
+        }
+        // for all the remaining names (c, d), create a new empty register, because the
+        // user might access the variable later
+        if names.len() > exprs.len() {
+            let reg_map = self.curr_reg_map();
+            for i in exprs.len()..names.len() {
+                reg_map.create_reg(names[i]);
+            }
+        } else if names.len() < exprs.len() {
+            // make sure we also compile every expression on the right side
+            // local a = 1, 2, f(); we have to also compile 2, and f()
+            for i in names.len()..exprs.len() {
+                self.compile_expr(exprs[i]);
+            }
+        }
+    }
+
+    /// Compiles a multi-assignemnt (a combination of local and global assignments).
+    /// * `names` - the variable names
+    /// * `exprs` - the expressions that are assigned
+    fn compile_assignments(&mut self, names: Vec<&'a str>, exprs: Vec<&'a Node<u8>>) {
+        // example: x, y, z, w = 1, 2
+        // compile x = 1, y = 2
+        for (name, expr) in names.iter().zip(exprs.iter()) {
+            self.compile_assignment(name, expr, false);
+        }
+        // for all the remaining names (z, w), create a new empty register, and update
+        // _ENV if the variable has not been declared as local in some outer scope
+        if names.len() > exprs.len() {
+            for i in exprs.len()..names.len() {
+                let reg = self.curr_reg_map().get_new_reg();
+                if !self.curr_reg_map().is_local(names[i]) {
+                    self.add_to_env(names[i], reg);
+                } else {
+                    self.curr_reg_map().set_reg(names[i], reg);
+                }
+            }
+        } else if names.len() < exprs.len() {
```

Comment:
What happens if `names.len() == exprs.len()`?

Reply:
It might be worth adding a comment to explain this: the above logic is a classic mistake, so it immediately rang my alarm bells.